connect-api/src/main/java/com/ifttt/connect/TokenInterceptor.java

```diff
@@ -19,14 +20,34 @@ void setToken(String token) {
         this.token = token;
     }
 
+    void setCredentialsProvider(@Nullable CredentialsProvider credentialsProvider) {
+        this.credentialsProvider = credentialsProvider;
+    }
+
     boolean isUserAuthenticated() {
         return token != null;
     }
 
     @Override
     public Response intercept(Chain chain) throws IOException {
         if (token == null) {
-            return chain.proceed(chain.request());
+            if (credentialsProvider != null) {
+                try {
+                    token = credentialsProvider.getUserToken();
+                    if (token == null) {
+                        // If token is still null, proceed without it.
+                        return chain.proceed(chain.request());
+                    }
+                } catch (Exception e) {
+                    if (BuildConfig.DEBUG) {
+                        e.printStackTrace();
+                    }
+
+                    return chain.proceed(chain.request());
```

Comment:
this statement can be outside the try-catch block since it is common amongst both.

Reply:
It is only called when the `token` is null in the try block. If it is not null, we are going to proceed with using the token below.

Reply:
ah yes.